### ISSUE TITLE
Copying an HDUList is shallow

### ISSUE THREAD

**Issue by pearsonkyle** [astropy/astropy#6395]
Just a suggestion here:

It might be worth adding a deep copy method to HDUList such that is behaves in the same manner as the copy method in an PrimaryHDU object. The parent class for an HDUList has a 'copy' method but it is only a shallow copy since it inherits from a list. 

It would be convenient to have these behaviors similar because when you read in a fits file it will automatically create an HDUList and sometimes you may want to copy it without reading the file in again.

**Comment by MSeifert04** [comment]
Does it work as expected if you `copy.deepcopy` it?

**Comment by pearsonkyle** [comment]
@MSeifert04 
I can not deepcopy it. A type error is arises and says "can't pickle mmap.mmap objects" 

This was with Python 3.6.0

**Comment by MSeifert04** [comment]
That's a tricky question because what would be the expected behavior of a copied `memmap`?

Does it work with `copy.deepcopy` if you specify `memmap=False` when opening it?

**Comment by pearsonkyle** [comment]
@MSeifert04 
Opening the file with `memmap=False` and then trying to `copy.deepcopy` still throws a TypeError however this time it says `TypeError: cannot serialize '_io.FileIO' object`

**Comment by MSeifert04** [comment]
Is that really a common use-case? My sense would be that just creating a new `HDUList` based on the current hdus would be better:

```
def copy(hdulist):
    return fits.HDUList([hdu.copy() for hdu in hdulist])
```

and then overwrite `HDUList.copy` to `raise NotImplementedError` because getting (and more importantly keeping) that to work properly across all cases would be really tricky especially if it needs to keep in sync with a fileobject (and all the race conditions that entails if one opens the file in update mode).

**Comment by MSeifert04** [comment]
cc: @saimn,@pllim

**Comment by saimn** [comment]
As HDU objects have a copy method, we could indeed add a (deep)copy to HDUList, with the implementation suggested by @MSeifert04 . I wonder though if it could be useful to keep `copy` to do a shallow copy, and add a `deepcopy` method. A usecase would be to replace only one HDU, with a shallow copy you can avoid doing a full copy of all HDUs.

**Comment by pllim** [comment]
I have no opinion on this but maybe @drdavella does?

**Comment by drdavella** [comment]
I'm not sure I have much to add at this point. It sounds like @MSeifert04 has suggested a pretty reasonable way to implement copy for `HDUList`. Can this proposed copy function just be used as the override of the copy method for `HDUList`? If the main concern is synchronization, is it enough to say that copy just represents a single snapshot in time so it doesn't necessarily have to worry about staying in sync with any underlying files or memory maps?

Also, if anything I just said implies that I have absolutely no understanding of what you're talking about, please feel free to say so.

**Comment by cdeil** [comment]
I have the following use case: I want to read a FITS file with tables like this one:
```
wget https://astropy.stsci.edu/data/tutorials/FITS-tables/chandra_events.fits -O data.fits
```
and create a new version of that file that is identical, except that I want to apply a row subset selection in one of the tables.

Note that this file has several tables:
```
>>> from astropy.io import fits
>>> fits.open('data.fits')
[<astropy.io.fits.hdu.image.PrimaryHDU object at 0x104564828>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x1045c2da0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x104ec6ef0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x104eeeb00>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x104ef8828>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x104f00550>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x104f05278>]
```

So I started my script with
```
hdu_list = fits.open('data.fits').copy()
```
but to my surprise there were no tables in `hdu_list`.

Check this out:
```
>>> from astropy.io import fits
>>> hdu_list = fits.open('data.fits')
>>> hdu_list.copy()
[<astropy.io.fits.hdu.image.PrimaryHDU object at 0x1075b7860>]
```
The copy is "loosing" the tables.

But it gets weirder. If you execute the HDUList repr, and call copy after, then it seems to work:
```
>>> from astropy.io import fits
>>> hdu_list = fits.open('data.fits')
>>> hdu_list
[<astropy.io.fits.hdu.image.PrimaryHDU object at 0x1121ea8d0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112248e48>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b4cf98>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b75ba8>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b798d0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b855f8>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b89320>]
>>> hdu_list.copy()
[<astropy.io.fits.hdu.image.PrimaryHDU object at 0x1121ea8d0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112248e48>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b4cf98>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b75ba8>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b798d0>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b855f8>, <astropy.io.fits.hdu.table.BinTableHDU object at 0x112b89320>]
```
I think the current behaviour of `HDUList.copy` is confusing at best, and could even lead to madness in some cases, no? (I know it was a close encounter for me just now.)

Probably this issue isn't easy to avoid, because `fits.open` doesn't leave the HDU list in a state where the number of HDUs is known because the file hasn't been scanned? And that is only triggered later, in this case by `repr`? Is this a good idea?

So +1 to overwrite `copy` as suggested, or if that would leave other methods still with a similar issue (unpredictable behaviour; depending on whether repr is called or not between `fits.open` and method call) to change `fits.open` (e.g. by adding a `repr(hdu_list)` or some other call to trigger the loading of info which HDUs are in the file at the end before `return hdu_list`).

**Comment by saimn** [comment]
That's probably because of the lazy loading, I guess `.copy` currently copies only the loaded HDUs. Could you with `fits.open('data.fits', lazy_load_hdus=False)` to be sure ?

**Comment by saimn** [comment]
Fixed by  #7218 (shallow copy with `.copy()` and deep with `copy.deepcopy`).
